S3 1.0 Breaking Change Documentation

johnny-schmidt opened the pull request:
Breaking change documentation and `metadata.yaml` changes for S3 1.0.

I will merge into the graphic stack for release with all the underlying breaking changes.

Previews (Rendered): 
- [New Docs](https://github.com/airbytehq/airbyte/blob/epic-8349-certify-s3/breaking-changes-and-metadata/docs/integrations/destinations/s3.md)
- [Migration Guide](https://github.com/airbytehq/airbyte/blob/epic-8349-certify-s3/breaking-changes-and-metadata/docs/integrations/destinations/s3-migrations.md)

johnny-schmidt commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/43374?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#43374** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/43374?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#42939** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/42939?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#42542** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/42542?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#42409** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/42409?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>: 1 other dependent PR ([#42579](https://github.com/airbytehq/airbyte/pull/42579) <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/42579?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>)
* **#42405** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/42405?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>: 2 other dependent PRs ([#42543](https://github.com/airbytehq/airbyte/pull/42543) <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/42543?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>, [#42924](https://github.com/airbytehq/airbyte/pull/42924) <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/42924?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>)
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @johnny-schmidt and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

edgao commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,202 @@
+|-------------------------|--------------------------|--------------------|-----------------------------------------------------------|----------------------------------------------|--------------------------------------------------------------------------------------------------------|
+| `_airbyte_ab_id`        | `_airbyte_raw_id`        | string (UUID)      | `{ "type": "string", "logicalType": "uuid" }`             | `String`                                     | Airbyte-added unique identifier.                                                                       |
+| `_airbyte_emitted_at`   | `_airbyte_extracted_at`  | integer (epoch ms) | `{ "type": "string", "logicalType": "timestamp-millis" }` | `Int64(logicalType=Timestamp(Milliseconds))` | Time at which the data was loaded from the source.                                                     |
+| [NEW FIELD]             | `_airbyte_generation_id` | integer            | `{ "type": "long" }`                                      | `Int64`                                      | Monotonically-increasing refresh id, if applicable.                                                    |
```
my initial glance missed the field renamings, b/c I went straight to the NEW FIELD rows. I think this would be a bit clearer at implying that the other rows also have important information

```suggestion
| [NONEXISTENT]             | `_airbyte_generation_id` | integer            | `{ "type": "long" }`                                      | `Int64`                                      | Monotonically-increasing refresh id, if applicable.                                                    |
```

johnny-schmidt replied:
Done

edgao commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,202 @@
+
+| Field Name     | Json Type | Description                                                    |
+|----------------|-----------|----------------------------------------------------------------|
+| `changes`      | list      | Change data capture info                                       |
```
let's not use the literal phrase "change data capture", since it's entirely unrelated to database CDC
```suggestion
| `changes`      | list      | Modifications to the data to accomodate it for storage in S3                                       |
```

evantahler replied:
+1, as "CDC" is a term-of-art in the data moving sector

johnny-schmidt replied:
```A record of any changes airbyte made to the data for compatibility and/or to handle errors```

edgao commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,202 @@
+
+| Field Name | Json Type* | Description                                                                |
+|------------|------------|----------------------------------------------------------------------------|
+| `field`    | string     | The name of the field that was affected                                    |
```
this is the jsonpath, right? or do we actually only capture this for top-level fields?

johnny-schmidt replied:
I can't remember. The path is passed into the listener, but I think I just use the field name. It's an easy fix though, I'll check.

johnny-schmidt replied:
Specifically this is how it behaves right now:
* any field failure at any depth triggers a capture
* however, only the field name is used

The reason for this was because it was a choice between the original name (which might be ambiguous if there nested objects with the same fields) and a path made of avro-mangled names.

The path is more informative though, and technically matches what they'd be looking at in their data lake, so it's probably the better choice?

johnny-schmidt replied:
Actually, I'm going to skip on this and pocket it as an improvement. The path passed out of the json-to-avro converter is not usable, because it doesn't account for array indexes. Fixing it is a major change.

johnny-schmidt replied:
https://app.zenhub.com/workspaces/destinations-6332532d6893ce001c7a56a7/issues/gh/airbytehq/airbyte/43397

edgao commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,202 @@
+|------------|------------|----------------------------------------------------------------------------|
+| `field`    | string     | The name of the field that was affected                                    |
+| `change`   | string     | The type of change (currently only `NULLED` or `TRUNCATED`)                |
+| `reason`   | string     | The reason for the change, including its origin (ie, destination-v-source) |
```
technically, platform could also be a change origin
```suggestion
| `reason`   | string     | The reason for the change, including its origin (e.g. source/destination) |
```

johnny-schmidt replied:
I know, I was trying to be less verbose. I added it in the s3.md though and will add it here as well.

johnny-schmidt replied:
fixed

edgao commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,202 @@
+| `{ "type": "object", "properties": { "id": { "type": "integer" } } }` | `{ "id": 1, "name": "Alice" }` | `{ "type": "object", "properties": { "id": { "type": "integer" } } }` | `{ "id": 1 }`                        |
+| `{ "type": "object" }`                                                | `{ "id": 1, "name": "Alice" }` | `{ "type": "string" }`                                                | `"{\"id\": 1, \"name\": \"Alice\"}"` |
+
+This behavior will be applied to both Avro and Parquet formats.
```
probably worth (a) moving this to right after the `Formerly, objects without schemas...` sentence, and (b) explicitly stating that json/csv output are unaffected

johnny-schmidt replied:
will do

johnny-schmidt replied:
`Formerly, in both Avro and Parquet formats, objects without schemas`

edgao commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,202 @@
+
+### Array types
+
+Formerly, arrays without types (json-schema with `"type": "array"` but no `items` field) were converted to native arrays of string serializations of the underlying types. The same behavior was applied to arrays of union types (`{ "type": "array", "items": { "oneOf": [ /* various types */ ] }`).
```
```suggestion
Formerly, arrays without types (`{"type": "array"}` with no `items` field) were converted to native arrays of string serializations of the underlying types. The same behavior was applied to arrays of union types (`{ "type": "array", "items": { "oneOf": [ /* various types */ ] }`).
```

johnny-schmidt replied:
will do

johnny-schmidt replied:
fixed

edgao commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,202 @@
+
+For example, the following input schemas and data formerly resulted in:
+
+| Input Schema                                                                           | Input Data     | Old Output Schema                                    | New Output Data  |
```
```suggestion
| Input Schema                                                                           | Input Data     | Old Output Schema                                    | Old Output Data  |
```

johnny-schmidt replied:
will do

johnny-schmidt replied:
fixed

edgao commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,202 @@
+| `{ "type": "array", "items": { "oneOf": [ {"type": "integer", "type": "string"} ] } }` | `[1, "Alice"]` | `{ "type": "array", "items": [ "null", "string" ] }` | `["1", "Alice"]` |
+| `{ "type": "array", "items": { "oneOf": [ {"type": "integer", "type": "string"} ] } }` | `[1, false]`   | `{ "type": "array", "items": [ "null", "string" ] }` | `["1", "false"]` |
+
+Now:
```
is this intentionally not combined with the previous table? (I'm guessing it's just too wide for comfortable reading?)

johnny-schmidt replied:
Yes just too wide

edgao commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,202 @@
+
+| Airbyte Union Type                                                                                                                     | New Parquet Type                                                                                                                               |
+|----------------------------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------|
+| `{"oneOf": [ {"type": "integer"}, {"type": "string"} ] }`                                                                              | `Optional Record { type: Optional String, integer: Optional Int32, string: Optional String }`                                                  |
```
should we clarify that `type` is always nonnull, even though it's generated as `Optional String`?

(... can we generate the schema as non-optional string?)

johnny-schmidt replied:
Yeah, that annoys me too. The existing schema converter makes everything nullable. There's no easy way to hack around it. I'll clarify.

johnny-schmidt replied:
I hacked around it. :) 


```
  }, {
    "name" : "union_with_dupes",
    "type" : [ "null", {
      "type" : "record",
      "name" : "union_with_dupes",
      "namespace" : "",
      "fields" : [ {
        "name" : "type",
        "type" : "string"
      }, {
        "name" : "integer",
        "type" : [ "null", "long" ],
        "default" : null
      }, {
        "name" : "string",
        "type" : [ "null", "string" ],
        "default" : null
      } ]
    } ],
    "default" : null
  } ]
```

But the hacks are in the old schema converter, so they can go away later.

I'll add tests for it and push the fix / update this documentation tomorrow morning.

edgao replied:
mrrgh. this is what I was afraid of, when we started passing around jsonschema blobs between mappers  :P  but similar to everything else, I think it goes away if we have a json/avro record converter that understands AirbyteType :shrug:

johnny-schmidt replied:
Tests added, documentation fixed

edgao commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,202 @@
+|-------------------------------|---------------------------------|-----------------------------------------------|
+| `{id: integer}`               | `{name: string}`                | `{id: integer, name: string}`                 |
+| `{id: integer, name: string}` | `{id: integer, birthday: date}` | `{id: integer, name: string, birthday: date}` |
+| `{id: integer}`               | `{id: string}`                  | [SYNC FAILED]                                 |
```
I missed this in the pr review - would it be hard to generate `{id: [integer, string]}` for this case?

johnny-schmidt replied:
Not hard, and an easy improvement. I *think* it's how the old code behaved, and I'd rather not add any more work at this point unless it's actually a regression. I'll double-check.

johnny-schmidt replied:
Ugh, I missed that this is what the old code actually did. Added.

evantahler commented on airbyte-integrations/connectors/destination-s3/metadata.yaml:
```diff
@@ -14,6 +14,12 @@ data:
+      1.0.0:
+        message: >
+          **This release includes breaking changes, including major revisions to the schema of stored data. Do not upgrade without reviewing the migration guide.**
+        upgradeDeadline: "2024-10-08"
```
👍 2 month breaking change window

evantahler commented on airbyte-integrations/connectors/destination-s3/metadata.yaml:
```diff
@@ -14,6 +14,12 @@ data:
+    breakingChanges:
+      1.0.0:
+        message: >
+          **This release includes breaking changes, including major revisions to the schema of stored data. Do not upgrade without reviewing the migration guide.**
```
👍 this is a fine change message, referencing other docs.

evantahler commented on docs/integrations/destinations/s3-migrations.md:
Check out your vercel preview (or run the doc server locally - check out the /docusourus dir's readme).  `*-migration` files aren't automatically added to the site.  You'll need to add a new sidebar block here: https://github.com/airbytehq/airbyte/blob/master/docusaurus/sidebars.js#L150-L164.  There are examples from other "Migration Guide" sections above

evantahler replied:
Also, this is a very detailed doc - well done!

johnny-schmidt replied:
<img width="265" alt="Screenshot 2024-08-08 at 3 18 26 PM" src="https://github.com/user-attachments/assets/57e76ca2-87cd-455e-81ac-906ccadad476">
done

evantahler commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,203 @@
+| Old field name          | New field name           | Json Type          | Avro Type                                                 | Parquet Type                                 | Description                                                                                            |
+|-------------------------|--------------------------|--------------------|-----------------------------------------------------------|----------------------------------------------|--------------------------------------------------------------------------------------------------------|
+| `_airbyte_ab_id`        | `_airbyte_raw_id`        | string (UUID)      | `{ "type": "string", "logicalType": "uuid" }`             | `String`                                     | Airbyte-added unique identifier.                                                                       |
+| `_airbyte_emitted_at`   | `_airbyte_extracted_at`  | integer (epoch ms) | `{ "type": "string", "logicalType": "timestamp-millis" }` | `Int64(logicalType=Timestamp(Milliseconds))` | Time at which the data was loaded from the source.                                                     |
```
```suggestion
| `_airbyte_emitted_at`   | `_airbyte_extracted_at`  | integer (epoch ms) | `{ "type": "string", "logicalType": "timestamp-millis" }` | `Int64(logicalType=Timestamp(Milliseconds))` | Time at which the data was extracted from the source.                                                     |
```

johnny-schmidt replied:
fixed

evantahler commented on docs/integrations/destinations/s3.md:
Above (Line ~310), it says "Incremental - Append + Deduped " is not supported - is that still true?

johnny-schmidt replied:
Deduping isn't supported in file destinations? Incremental append is though.

evantahler commented on docs/integrations/destinations/s3.md:
```diff
@@ -377,12 +377,29 @@ Like most of the other Airbyte destination connectors, usually the output has th
+| Column                   | Condition                                                                                         | Description                                                              |
+|:-------------------------|:--------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------|
+| `_airbyte_raw_id`        | Always exists                                                                                     | A uuid assigned by Airbyte to each processed record.                     |
+| `_airbyte_extracted_at`  | Always exists.                                                                                    | A timestamp representing when the event was pulled from the data source. |
```
[nit]

```suggestion
| `_airbyte_extracted_at`  | Always exists.                                                                                    | A timestamp representing when the event was loaded from the data source. |
```

Might as well use the term of art (Extract and Load)

johnny-schmidt replied:
fixed

katmarkham commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,203 @@
+
+## Data Changes
+
+This version introduces changes to the data types used when writing to S3 in avro or parquet. The changes are as follows:
```
```suggestion
This version introduces changes to the data types used when writing to S3 in Avro or Parquet. The changes are as follows:
```

katmarkham commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,203 @@
+
+### Date and time types
+
+This change introduces [this requested improvement](https://github.com/airbytehq/airbyte-internal-issues/issues/8973). Date and time types were stored in avro as unions of integral logical types and strings, which in parquet were represented as disjoint records. In practice, the date and time types were always converted, making the unions redundant. The resulting disjoint records were confusing.
```
```suggestion
This change introduces [this requested improvement](https://github.com/airbytehq/airbyte-internal-issues/issues/8973). Date and time types were stored in Avro as unions of integral logical types and strings, which in Parquet were represented as disjoint records. In practice, the date and time types were always converted, making the unions redundant. The resulting disjoint records were confusing.
```

katmarkham replied:
can you say something other than "this requested improvement"; be more specific.

johnny-schmidt replied:
`This change introduces [simplification of the handling of dates, times, and timestamps in Parquet and Avro]`?

katmarkham commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,203 @@
+| timestamp without timezone    | `Optional Record { member0: Optional Int64(logical_type=Timestamp(Microseconds)), member1: Optional String }` | `Optional Int64(logical_type=Timestamp(Microseconds))` |
+| timestamp with timezone       | `Optional Record { member0: Optional Int64(logical_type=Timestamp(Microseconds)), member1: Optional String }` | `Optional Int64(logical_type=Timestamp(Microseconds))` |
+
+Note: the times and timestamps with timezones are converted to UTC. In parquet, the field metadata `is_adjusted_to_utc` will always be `true`.
```
```suggestion
Note: the times and timestamps with timezones are converted to UTC. In Parquet, the field metadata `is_adjusted_to_utc` will always be `true`.
```

katmarkham commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,203 @@
+
+This version introduces changes to the data types used when writing to S3 in avro or parquet. The changes are as follows:
+
+### Primitive types
```
```suggestion
### Primitive Types
```

katmarkham commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,203 @@
+| integer                   | `["null", "long"]`       | `Optional Int64`      | [UNCHANGED]     | [UNCHANGED]      |
+| number                    | `["null", "double"]`     | `Optional Double`     | [UNCHANGED]     | [UNCHANGED]      |
+
+### Date and time types
```
```suggestion
### Date and Time Types
```

katmarkham commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,203 @@
+
+Note: the times and timestamps with timezones are converted to UTC. In parquet, the field metadata `is_adjusted_to_utc` will always be `true`.
+
+#### Time of day conversion bugfix
```
```suggestion
#### Time of Day Conversion- Bugfix
```

katmarkham commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,203 @@
+
+#### Time of day conversion bugfix
+
+Formerly, for the airbyte type `time_of_day_with_timezone`, [timezones were not respected when converting](https://github.com/airbytehq/airbyte/issues/43019). This has been fixed.
```
```suggestion
Formerly, for the Airbyte type `time_of_day_with_timezone`, [timezones were not respected when converting](https://github.com/airbytehq/airbyte/issues/43019). This has been fixed.
```

katmarkham replied:
when converting what?

katmarkham commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,203 @@
+
+### Object types
+
+Formerly, objects without schemas (json-schema with `"type": "object"` but no `properties` field), and properties not listed in the `properties` field, were accumulated in `_airbyte_additional_properties`. The new behavior is
```
```suggestion
Formerly, objects without schemas (json-schema with `"type": "object"` but no `properties` field), and properties not listed in the `properties` field, were accumulated in `_airbyte_additional_properties`. The new expected behavior is as follows:
```

katmarkham replied:
*JSON

katmarkham commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,203 @@
+
+Formerly, objects without schemas (json-schema with `"type": "object"` but no `properties` field), and properties not listed in the `properties` field, were accumulated in `_airbyte_additional_properties`. The new behavior is
+
+* undocumented fields in object with schemas are silently dropped
```
```suggestion
* Undocumented fields in objects with schemas are silently dropped.
```

katmarkham commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,203 @@
+Formerly, objects without schemas (json-schema with `"type": "object"` but no `properties` field), and properties not listed in the `properties` field, were accumulated in `_airbyte_additional_properties`. The new behavior is
+
+* undocumented fields in object with schemas are silently dropped
+* objects without schemas are serialized into a json string
```
```suggestion
* Objects without schemas are serialized into a JSON string.
```

katmarkham commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,203 @@
+
+* undocumented fields in object with schemas are silently dropped
+* objects without schemas are serialized into a json string
+* `_airbyte_additional_properties` is dropped entirely
```
```suggestion
* `_airbyte_additional_properties` is dropped entirely.
```

katmarkham commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,203 @@
+
+This behavior will be applied to both Avro and Parquet formats.
+
+Note: dropped fields will not appear in `_airbyte_meta.changes[]`. Additionally, n object with null or empty properties (`"properties": {}`) will be treated as a schemaless object. This is because this usually indicates an upstream source is failing to report its schema properly, and the data is not actually extraneous.
```
```suggestion
Note: dropped fields will not appear in `_airbyte_meta.changes[]`. Additionally, any object with null or empty properties (`"properties": {}`) will be treated as a schemaless object. This usually indicates an upstream source is failing to report its schema properly, and the data is not actually extraneous.
```

katmarkham commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,203 @@
+Note: dropped fields will not appear in `_airbyte_meta.changes[]`. Additionally, n object with null or empty properties (`"properties": {}`) will be treated as a schemaless object. This is because this usually indicates an upstream source is failing to report its schema properly, and the data is not actually extraneous.
+
+
+### Array types
```
```suggestion
### Array Types
```

katmarkham commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,203 @@
+
+### Array types
+
+Formerly, arrays without types (json-schema with `"type": "array"` but no `items` field) were converted to native arrays of string serializations of the underlying types. The same behavior was applied to arrays of union types (`{ "type": "array", "items": { "oneOf": [ /* various types */ ] }`).
```
```suggestion
Formerly, arrays without types (JSON-schema with `"type": "array"` but no `items` field) were converted to native arrays of string serializations of the underlying types. The same behavior was applied to arrays of union types (`{ "type": "array", "items": { "oneOf": [ /* various types */ ] }`).
```

katmarkham commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,203 @@
+
+Formerly, arrays without types (json-schema with `"type": "array"` but no `items` field) were converted to native arrays of string serializations of the underlying types. The same behavior was applied to arrays of union types (`{ "type": "array", "items": { "oneOf": [ /* various types */ ] }`).
+
+Now:
```
```suggestion
Now you should expect the following behavior from arrays:
```

katmarkham commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,203 @@
+
+Now:
+
+* arrays without types are serialized into json array strings
```
```suggestion
* Arrays without types are serialized into JSON array strings.
```

katmarkham commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,203 @@
+Now:
+
+* arrays without types are serialized into json array strings
+* arrays of unions are treated as arrays of mixed types
```
```suggestion
* Arrays of unions are treated as arrays of mixed types.
```

katmarkham commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,203 @@
+* arrays without types are serialized into json array strings
+* arrays of unions are treated as arrays of mixed types
+
+For example, the following input schemas and data formerly resulted in:
```
```suggestion
For example, let's compare the following input schemas and data to the new expectations:
```

katmarkham commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,203 @@
+
+### Union Types (Parquet Only)
+
+#### Disjoint record improvements
```
```suggestion
#### Disjoint Record Improvements
```

katmarkham commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,203 @@
+
+#### Disjoint record improvements
+
+Formerly, unions in parquet were represented as anonymous disjoint records. For example:
```
```suggestion
Formerly, unions in Parquet were represented as anonymous disjoint records. For example:
```

katmarkham commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,203 @@
+| `{"oneOf": [ {"type": "boolean"}, {"type": "object", "properties": { "id": { "type": "integer" }, "name": { "type": string" } } } ] }` | `Optional Record { type: Optional String, boolean: Optional Boolean, object: Optional Record { id: Optional Uint32, name: Optional String } }` |
+
+Where
+* only one field will always be set, and
```
```suggestion
* Only one field will always be set.
```

katmarkham commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,203 @@
+
+Where
+* only one field will always be set, and
+* the value of the `type` field will always be equal the name of the set field
```
```suggestion
* The value of the `type` field will always be equal the name of the `set `field.
```

johnny-schmidt replied:
Not sure about putting ` around "set" -- there is not a field called set. `"the field that was set"` instead?

katmarkham commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,203 @@
+| `1`                           | `{ "type": "integer", "integer": 1 }`                           |
+| `{"id": 10, "name": "Alice"}` | `{ "type": "object", "object": { "id": 10, "name": "Alice" } }` |
+
+#### Disjoint record type names
```
```suggestion
#### Disjoint Record-Type Names
```

johnny-schmidt replied:
Capitalization yes. It's "type names" for records, not the name of record types, so I'm not sure the hyphen makes sense.

katmarkham commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,203 @@
+| `{"type": "string", "format": "date-time", "airbyte_type": "timestamp_with_timezone" }`    | `Optional Int64(logical_type=Timestamp(Microseconds)))`                | `timestamp_with_timezone`    |
+| `{"type": "string", "format": "date-time", "airbyte_type": "timestamp_without_timezone" }` | `Optional Int64(Optional Int64(logical_type=Timestamp(Microseconds)))` | `timestamp_without_timezone` |
+
+#### Merging like options in unions
```
```suggestion
#### Merging Like Options in Unions
```

katmarkham commented on docs/integrations/destinations/s3.md:
```diff
@@ -377,12 +377,29 @@ Like most of the other Airbyte destination connectors, usually the output has th
+|:-------------------------|:--------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------|
+| `_airbyte_raw_id`        | Always exists                                                                                     | A uuid assigned by Airbyte to each processed record.                     |
+| `_airbyte_extracted_at`  | Always exists.                                                                                    | A timestamp representing when the event was pulled from the data source. |
+| `_airbyte_generation_id` | Always exists.                                                                                    | An integer id that increases with each new refresh.                      | 
```
```suggestion
| `_airbyte_generation_id` | Always exists                                                                                    | An integer id that increases with each new refresh.                      | 
```

katmarkham commented on docs/integrations/destinations/s3.md:
```diff
@@ -377,12 +377,29 @@ Like most of the other Airbyte destination connectors, usually the output has th
+| `_airbyte_raw_id`        | Always exists                                                                                     | A uuid assigned by Airbyte to each processed record.                     |
+| `_airbyte_extracted_at`  | Always exists.                                                                                    | A timestamp representing when the event was pulled from the data source. |
+| `_airbyte_generation_id` | Always exists.                                                                                    | An integer id that increases with each new refresh.                      | 
+| `_airbyte_meta`          | Always exists.                                                                                    | A structured object containing metadata about the record.                |
```
```suggestion
| `_airbyte_meta`          | Always exists                                                                                    | A structured object containing metadata about the record.                |
```

katmarkham commented on docs/integrations/destinations/s3.md:
```diff
@@ -377,12 +377,29 @@ Like most of the other Airbyte destination connectors, usually the output has th
+| `_airbyte_extracted_at`  | Always exists.                                                                                    | A timestamp representing when the event was pulled from the data source. |
+| `_airbyte_generation_id` | Always exists.                                                                                    | An integer id that increases with each new refresh.                      | 
+| `_airbyte_meta`          | Always exists.                                                                                    | A structured object containing metadata about the record.                |
+| `_airbyte_data`          | When no normalization \(flattening\) is needed, all data reside under this column as a json blob. |                                                                          |
```
```suggestion
| `_airbyte_data`          | When no normalization \(flattening\) is needed, all data resides under this column as a JSON blob. |                                                                          |
```

katmarkham commented on docs/integrations/destinations/s3.md:
```diff
@@ -377,12 +377,29 @@ Like most of the other Airbyte destination connectors, usually the output has th
+| `_airbyte_data`          | When no normalization \(flattening\) is needed, all data reside under this column as a json blob. |                                                                          |
+| root level fields        | When root level normalization \(flattening\) is selected, the root level fields are expanded.     |                                                                          |
+
+The schema for airbyte meta is:
```
```suggestion
The schema for Airbyte meta is:
```

katmarkham replied:
should this be underscored?

johnny-schmidt replied:
Should be `_airbyte_meta`

katmarkham commented on docs/understanding-airbyte/json-avro-conversion.md:
```diff
@@ -348,7 +354,7 @@ Note that the first object in `array_field` originally does not have a `message`
 ### Untyped Array
 
-When a Json array field has no `items`, the element in that array field may have any type. However, Avro requires that each array has a clear type specification. To solve this problem, the elements in the array are forced to be `string`s.
+When a Json array field has no `items`, the element in that array field may have any type. However, Avro requires that each array has a clear type specification. To solve this problem, the json array is serialized into its string representation.
```
```suggestion
When a JSON array field has no `items`, the element in that array field may have any type. However, Avro requires that each array has a clear type specification. To solve this problem, the JSON array is serialized into its string representation.
```

katmarkham commented on docs/understanding-airbyte/json-avro-conversion.md:
````diff
@@ -392,33 +395,24 @@ the corresponding Avro schema and object will be:
-  "default": null
-}
-```
+A Json object can have additional properties of unknown types, which is not compatible with the Avro schema. These properties will be silently dropped.
````
```suggestion
A JSON object can have additional properties of unknown types, which is not compatible with the Avro schema. These properties will be silently dropped.
```

katmarkham replied:
```suggestion
A JSON object can have additional properties of unknown types, which is not compatible with the Avro schema. These properties will be silently dropped.
```

katmarkham commented on docs/understanding-airbyte/json-avro-conversion.md:
```diff
@@ -453,20 +447,13 @@ will be converted to the following Avro object:
 ### Untyped Object
 
-If an `object` field has no `properties` specification, all fields within this `object` will be put into the aforementioned `_airbyte_additional_properties` field.
+If an `object` field has no `properties` specification, the entire json object will be serialized into its string representation.
```
```suggestion
If an `object` field has no `properties` specification, the entire JSON object will be serialized into its string representation.
```

katmarkham commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,203 @@
+
+This version introduces changes to the schema of data written to S3, which make it isomorphic to our [V2 certified database destinations](https://docs.airbyte.com/release_notes/upgrading_to_destinations_v2/#what-is-destinations-v2). The schema changes are as follows:
+
+| Old field name        | New field name           | Json Type          | Avro Type                                                 | Parquet Type                                 | Description                                                                                            |
```
```suggestion
| Old field name        | New field name           | JSON Type          | Avro Type                                                 | Parquet Type                                 | Description                                                                                            |
```

katmarkham commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,203 @@
+
+The `_airbyte_meta` field is an object that currently has one field:
+
+| Field Name     | Json Type | Description                                                                                |
```
```suggestion
| Field Name     | JSON Type | Description                                                                                |
```

katmarkham commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,203 @@
+
+The `changes` field is a list of objects, each of which represents a change to the data. Each object has the following fields:
+
+| Field Name | Json Type* | Description                                                                        |
```
```suggestion
| Field Name | JSON Type* | Description                                                                        |
```

katmarkham commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,203 @@
+
+All primitive types are unchanged:
+
+| Airbyte Json Schema Type  | Old Avro Type            | Old Parquet Type      | New Avro Type   | New Parquet Type |
```
```suggestion
| Airbyte JSON Schema Type  | Old Avro Type            | Old Parquet Type      | New Avro Type   | New Parquet Type |
```

katmarkham commented on docs/integrations/destinations/s3.md:
```diff
@@ -377,12 +377,29 @@ Like most of the other Airbyte destination connectors, usually the output has th
+| `change`   | string | The type of change (eg, `NULLED`, `TRUNCATED`).                                                                          |
+| `reason`   | string | The reason for the change, including its system of origin (ie, whether it was a source, destination, or platform error). |
 
 For example, given the following json object from a source:
```
```suggestion
For example, given the following JSON object from a source:
```

katmarkham commented on docs/understanding-airbyte/json-avro-conversion.md:
```diff
@@ -348,7 +354,7 @@ Note that the first object in `array_field` originally does not have a `message`
-When a Json array field has no `items`, the element in that array field may have any type. However, Avro requires that each array has a clear type specification. To solve this problem, the elements in the array are forced to be `string`s.
+When a Json array field has no `items`, the element in that array field may have any type. However, Avro requires that each array has a clear type specification. To solve this problem, the json array is serialized into its string representation.
 
 For example, given the following Json schema and object:
```
```suggestion
For example, given the following JSON schema and object:
```

katmarkham commented on docs/understanding-airbyte/json-avro-conversion.md:
````diff
@@ -392,33 +395,24 @@ the corresponding Avro schema and object will be:
-```
+A Json object can have additional properties of unknown types, which is not compatible with the Avro schema. These properties will be silently dropped.
 
 For example, given the following Json schema:
````
```suggestion
For example, given the following JSON schema:
```

katmarkham commented on docs/understanding-airbyte/json-avro-conversion.md:
```diff
@@ -453,20 +447,13 @@ will be converted to the following Avro object:
-If an `object` field has no `properties` specification, all fields within this `object` will be put into the aforementioned `_airbyte_additional_properties` field.
+If an `object` field has no `properties` specification, the entire json object will be serialized into its string representation.
 
 For example, given the following Json schema and object:
```
```suggestion
For example, given the following JSON schema and object:
```

edgao reviewed (approved):
lgtm!

johnny-schmidt commented:
@katmarkham I took all your suggestions except where noted. I also fixed all of the remaining of:

* title-cased heading names
* json to JSON throughout
* turned all bullet list items into sentences (capitalized and punctuated)
* capitalized Avro and Parquet throughout

johnny-schmidt commented:
/publish-java-cdk

aaronsteers reviewed (approved):
I think this looks really good. My only high-level feedback is that (if time allows), the data type mapping logic could be helpful on the main docs page. There's a lot of detail in the migration guide which for new users would likely be helpful in their upfront setup. With an eye to not duplicating much of the same docs, I wonder if it makes sense to add a section under Parquet in the docs, that links to the relevant section in the migration doc?

That said, I don't think that needs to block this PR so I'm happy to approve as-is.

johnny-schmidt commented:
Pardon the code changes here. That was supposed to go with the underlying PR -- fix for something that came up during prod testing

johnny-schmidt commented:
> I think this looks really good. My only high-level feedback is that (if time allows), the data type mapping logic could be helpful on the main docs page. There's a lot of detail in the migration guide which for new users would likely be helpful in their upfront setup. With an eye to not duplicating much of the same docs, I wonder if it makes sense to add a section under Parquet in the docs, that links to the relevant section in the migration doc?
> 
> That said, I don't think that needs to block this PR so I'm happy to approve as-is.

Yes I called out writing a Parquet doc as a follow-up.

katmarkham commented on docs/integrations/destinations/s3-migrations.md:
```diff
@@ -0,0 +1,202 @@
+
+#### Disjoint Record Type Names
+
+The following type and field names will be used for each airbyte type:
```
```suggestion
The following type and field names will be used for each Airbyte type:
```

johnny-schmidt commented:
@katmarkham 
* fixed the remaining a/Airbyte
* replaced all Json with JSON in the old doc
* added this language for why the change is good

```This version introduces changes to the schema of data written to S3, which make it isomorphic to our [V2 certified database destinations](https://docs.airbyte.com/release_notes/upgrading_to_destinations_v2/#what-is-destinations-v2), as well as various improvements to format conversion.

* Conversion failures are captured: values not matching the client schema will no longer break syncs using Avro or Parquet formats.
* Changes introduced by the source or platform, such as NULLing bad values or truncating long ones, are visible in the metadata.
* There is improved handling of various types in Avro and Parquet formats, including simplification of date and time types, more human-readable schemaless objects and arrays, and better support for unions in Parquet.
* Sync and generation ids are made available, providing more information for debugging.```

katmarkham commented on airbyte-integrations/connectors/destination-s3/metadata.yaml:
```diff
@@ -14,12 +14,18 @@ data:
+    breakingChanges:
+      1.0.0:
+        message: >
+          **This release includes breaking changes, including major revisions to the schema of stored data. Do not upgrade without reviewing the migration guide.**
```
```suggestion
          **This version introduces changes to the schema of data written to S3, which make it isomorphic to our [V2 certified database destinations](https://docs.airbyte.com/release_notes/upgrading_to_destinations_v2/#what-is-destinations-v2), as well as various improvements to format conversion. This release includes breaking changes, including major revisions to the schema of stored data. Do not upgrade without reviewing the migration guide.**
```

johnny-schmidt commented:
Graphite folded this into the underlying PR without merging this one. Closing it manually.
